srai/regionizers/h3_regionizer.py

```diff
@@ -0,0 +1,191 @@
+"""
+H3 regionizer.
+
+This module exposes Uber's H3 Hexagonal Hierarchical Geospatial Indexing System [1] as a regionizer.
+
+Note:
+    The default API [2] was chosen (basic_str) to ease the implementation.
+    It may be beneficial to try the NumPy API for computationally-heavy work.
+
+References:
+    [1] https://uber.github.io/h3-py/
+    [2] https://uber.github.io/h3-py/api_comparison
+
+"""
+
+from typing import List
+
+import geopandas as gpd
+import h3
+import pandas as pd
+from functional import seq
+from shapely import geometry
+
+
+class H3Regionizer:
+    """
+    H3 Regionizer.
+
+    H3 Regionizer allows the given geometries to be divided
+    into H3 cells - hexagons with pentagons as a very rare exception
+
+    """
+
+    def __init__(self, resolution: int, buffer: bool = True) -> None:
+        """
+        Init H3Regionizer.
+
+        Args:
+            resolution (int): Size of the cells.
+            buffer (bool, optional): Whether to fully cover the geometries with
+                H3 Cells (visible on the borders).
+                Turn off for large geometries, as it's computationally expensive.
+                Defaults to True.
+
+        Raises:
+            ValueError: If resolution is not within 0 <= resolution <= 15.
+
+        """
+        if not (0 <= resolution <= 15):
+            raise ValueError(f"Resolution {resolution} is not within 0 <= resolution <= 15.")
+
+        self.resolution = resolution
+        self.buffer = buffer
+
+    def transform(self, gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```

Comment:
Logically, all required functions should be placed above this function. If it wasn't inside a class, it wouldn't compile because of it.
I'd put `_buffer`, `_gdf_from_h3_indexes`, `_h3_index_to_shapely_polygon` and the rest in the correct order above the `transform` function.

Reply:
I think I have another opinion on that. I want the class to be read clearly from top to bottom. Starting with what's the most important (the public API - transform()) and then the implementation details in order of invocation. I understand the logical aspect, that something must be first declared to be used later, but it is not the way we read. I would follow the [stepdown rule](https://marekhudyma.com/code-style/2021/03/02/step-down-rule.html) by the well-known author that shares the surname with the GOT writer xD

Reply:
Alright, I can accept that.